src/ringmesh/basic/attributes.cpp

```diff
@@ -0,0 +1,550 @@
+        attributes_.erase( it );
+    }
+
+    void AttributesManager::delete_attribute_store( AttributeStore* as )
+    {
+        for( auto& it : attributes_ )
+        {
+            if( it.second.get() == as )
+            {
+                attributes_.erase( it.first );
+                return;
+            }
+        }
+        ringmesh_assert_not_reached;
+    }
+
+    void AttributesManager::clear( bool keep_attributes )
+    {
+        if( keep_attributes )
+        {
+            for( auto& it : attributes_ )
+            {
+                it.second->clear();
+            }
+        }
+        else
+        {
+            attributes_.clear();
+        }
+        nb_items_ = 0;
+    }
+
+    void AttributesManager::copy( const AttributesManager& rhs )
+    {
+        clear( false );
+        resize( rhs.nb_items() );
+        for( auto& it : attributes_ )
+        {
+            bind_attribute_store( it.first, it.second->clone() );
+        }
+    }
+
+    /************************************************************************/
+
+    std::string ReadOnlyScalarAttributeAdapter::attribute_base_name(
+        const std::string& name )
+    {
+        size_t pos = name.find( '[' );
+        if( pos == std::string::npos )
+        {
+            return name;
+        }
+        return name.substr( 0, pos );
+    }
+
+    index_t ReadOnlyScalarAttributeAdapter::attribute_element_index(
+        const std::string& name )
+    {
+        index_t result = 0;
+        size_t pos = name.find( '[' );
+        if( pos != std::string::npos )
+        {
+            try
+            {
+                if( pos + 2 > name.length() )
+                {
+                    result = index_t( -1 );
+                }
+                else
+                {
+                    result = GEO::String::to_uint(
+                        name.substr( pos + 1, name.length() - pos - 2 ) );
+                }
+            }
+            catch( ... )
+            {
+                result = index_t( -1 );
+            }
+        }
+        return result;
+    }
+
+    void ReadOnlyScalarAttributeAdapter::bind_if_is_defined(
+        const AttributesManager& manager, const std::string& name )
+    {
+        ringmesh_assert( !is_bound() );
+        manager_ = &manager;
+        element_index_ = attribute_element_index( name );
+        store_ = manager_->find_attribute_store( attribute_base_name( name ) );
+
+        if( store_ == nil || element_index_ == index_t( -1 ) )
+        {
+            store_ = nil;
+            element_index_ = index_t( -1 );
+            return;
+        }
+
+        // Test element_index_ validity: should be smaller than
+        // store's dimension (or 2*store dimension if a vec2,
+        // or 3*store's dimension if a vec3)
+        if( element_index_ >= nb_scalar_elements_per_item() )
+        {
+            store_ = nil;
+            element_index_ = index_t( -1 );
+            return;

```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this redundant jump. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3626)